File: sdk/src/main/java/com/meniga/sdk/webservices/requests/UpdateChallenge.java
package com.meniga.sdk.webservices.requests;

import com.meniga.sdk.helpers.MenigaDecimal;

import org.joda.time.DateTime;

import java.util.List;
import java.util.UUID;

/**
 * Copyright 2017 Meniga Iceland Inc.
 */
public class UpdateChallenge extends QueryRequestObject {
	public transient UUID id;
	public String title;
	public String description;
	public DateTime startDate;
	public DateTime endDate;
	public String iconUrl;
	public TypeData typeData;

    @Override
	public long getValueHash() {
	    int result = id != null ? id.hashCode() : 0;
	    result = 31 * result + (title != null ? title.hashCode() : 0);
	    result = 31 * result + (description != null ? description.hashCode() : 0);
	    result = 31 * result + (startDate != null ? startDate.hashCode() : 0);
	    result = 31 * result + (endDate != null ? endDate.hashCode() : 0);
	    result = 31 * result + (iconUrl != null ? iconUrl.hashCode() : 0);
	    result = 31 * result + (typeData != null ? typeData.hashCode() : 0);
	    return result;
	}

	public static class TypeData {
		public MenigaDecimal targetAmount;
		public List<Long> categoryIds;
		public String metaData;

		@Override
		public int hashCode() {
			int result = targetAmount != null ? targetAmount.hashCode() : 0;
			result = 31 * result + (categoryIds != null ? categoryIds.hashCode() : 0);
			result = 31 * result + (metaData != null ? metaData.hashCode() : 0);
			return result;
		}
	}
}

File: sdk/src/main/java/com/meniga/sdk/webservices/challenge/UpdateChallenge.kt
/*
 * Copyright 2017-2018 Meniga Iceland Inc.
 */
package com.meniga.sdk.webservices.challenge

import com.meniga.sdk.helpers.MenigaDecimal
import com.meniga.sdk.webservices.requests.QueryRequestObject

import org.joda.time.DateTime
import java.util.UUID

internal data class UpdateChallenge(
        @JvmField var id: UUID? = null,
        @JvmField var title: String? = null,
        @JvmField var description: String? = null,
        @JvmField var startDate: DateTime? = null,
        @JvmField var endDate: DateTime? = null,
        @JvmField var iconUrl: String? = null,
        @JvmField var typeData: TypeData? = null
) : QueryRequestObject() {

    override fun getValueHash(): Long = hashCode().toLong()

    data class TypeData(
            @JvmField var targetAmount: MenigaDecimal? = null,
            @JvmField var categoryIds: List<Long>? = null,
            @JvmField var metaData: String? = null
    )
}
